### ISSUE TITLE
feat: mark NEXTAUTH_URL as optional in environment variable schema

### ISSUE THREAD

**Issue by AndKenneth** [t3-oss/create-t3-app#535]
### Is your feature request related to a problem? Please describe.

The default behavior of next auth is to fallback to VERCEL_URL when NEXTAUTH_URL is unavailable.
https://next-auth.js.org/configuration/options#nextauth_url

Marking NEXAUTH_URL as optional will prevent the need to set NEXTAUTH_URL when deploying to Vercel.

The new dev experience debugging the schema errors would generally lead someone adding the primary application URL to NEXTAUTH_URL, which means they need to juggle env vars for different environments and lose the benefits of the provided VERCEL_URL variable.

### Describe the solution you'd like to see

``` ts
export const serverSchema = z.object({
...
  NEXTAUTH_URL: z.string().url().optional(),
});
```

### Desribe alternate solutions

Could instead add a comment suggesting the use of .optional if the plan is to deploy to vercel. This would make it slightly less opinionated.

It also may be possible to check the presence of either VERCEL_URL or NEXTAUTH_URL, but I'm not sure how to do that myself.

### Additional information

It may also be a good idea to add a comment to the example .env file explaining that it does not need to be set on Vercel, but may need to be on other hosting platforms.

**Comment by Samathingamajig** [comment]
When set to `.optional()`, if someone clears it in `.env` there will also be issues when developing locally. Not a big issue but it doesn't seem like this was mentioned already in the big thread.

**Comment by TheoBr** [comment]
Maybe make vercel_url OR nextauth_url mandatory? Either way this 100% should be a thing as it's confused a ton of users in the discord.

**Comment by AndKenneth** [comment]
> When set to `.optional()`, if someone clears it in `.env` there will also be issues when developing locally. Not a big issue but it doesn't seem like this was mentioned already in the big thread.

Considering that localhost:3000 is the default for localhost I don't think this would really come up for solo devs as someone would have to specifically delete it. However it *could* also happen when a new dev is being setup on an existing project, because the .env won't exist.

**Comment by Samathingamajig** [comment]
I just realized that Next auth will also fail with the current default `NEXTAUTH_URL` if port 3000 is already taken up, because if 3000 is already taken then Next automatically takes the next port (usually 3001) so if it's anything other than 3000 it will break

**Comment by juliusmarminge** [comment]
Duplicate #216

**Comment by c-ehrlich** [comment]
Thought about it again and I think I was wrong about this in #216. IMO we should make it optional, but explain clearly in a comment why it is this way and link to relevant documentation.

**Comment by juliusmarminge** [comment]
Suggestion, add this as a comment or in docs:
```ts
export const serverSchema = z.object({
  ...
  // NEXTAUTH_URL: z.string().url().default(process.env.VERCEL_URL),
  NEXTAUTH_URL: z.string().url(),
});
```
Should still fail in prod if you're not on Vercel and there you'd have to supply your own?

Testing if this works now

**Comment by juliusmarminge** [comment]
This seems like it works, could need a bit more testing
```ts
  NEXTAUTH_URL: z.preprocess(
    (str) => process.env.VERCEL_URL ?? str,
    z.string().url()
  ),
```

Bringing in the sod-expert @KATT. Any suggestions?

**Comment by KATT** [comment]
> This seems like it works, could need a bit more testing
> 
> ```ts
>   NEXTAUTH_URL: z.preprocess(
>     (str) => process.env.VERCEL_URL ?? str,
>     z.string().url()
>   ),
> ```
> 
> Bringing in the zod-expert @KATT. Any suggestions?
> 
> Want the NEXTAUTH_URL to be required but if the VERCEL_URL is there it should use that.


LGTM.

Some random takes unrelated to this issue but on my `env.js` that might be useful:

- I don't really used `preprocess` much but seems decent here.
- I try to have my env vars parsing stupid
- I try to mostly avoid default values and optionality as I prefer the behavior to be explicit over implicit.
- `.transform`s I use a lot for e.g. booleans.
- `NEXTAUTH_URL` is really annoying and I wish they had a config where you could derive it from another env var instead of relying `process.env` like `const nextAuth = setupNextAuth({ baseUrl: env.APP_URL });`
- A consistent exception I have for env optionality is feature flagging 


Feature flag schema for ref:
```ts
const featureFlagSchema =  z.enum(['1', '0']).default('0').transform(v === '1');

const envSchema = z.object({
  FEATURE_COMMENT_FUNCTIONALITY_ENABLED: featureFlagSchema,

});
```

Maybe you should do a PR/issue to next-auth so the behavior could be explicit rather than depending on `process.env` globals

**Comment by bachitterch** [comment]
> Maybe we should do a PR/issue to next-auth so the behavior could be explicit rather than depending on `process.env` globals

Agree on that they have it for NEXTAUTH_SECRET so why for URL too,

**Comment by dtinth** [comment]
Workaround:

```ts
  // NEXTAUTH_URL is optional when VERCEL_URL is present.
  // See: https://next-auth.js.org/configuration/options#nextauth_url
  NEXTAUTH_URL: process.env.VERCEL_URL
    ? z.string().url().optional()
    : z.string().url(),
```

This gets inferred in `_serverEnv.data` as:

```
    NEXTAUTH_URL?: string | undefined;
```

**Comment by juliusmarminge** [comment]
> Workaround:
> 
> 
> 
> ```ts
> 
>   // NEXTAUTH_URL is optional when VERCEL_URL is present.
> 
>   // See: https://next-auth.js.org/configuration/options#nextauth_url
> 
>   NEXTAUTH_URL: process.env.VERCEL_URL
> 
>     ? z.string().url().optional()
> 
>     : z.string().url(),
> 
> ```
> 
> 
> 
> This gets inferred in `_serverEnv.data` as:
> 
> 
> 
> ```
> 
>     NEXTAUTH_URL?: string | undefined;
> 
> ```

That's incorrect. It should always be a string, just defaulted to VERCEL_URL if it exists

**Comment by c-ehrlich** [comment]
closing because #557 has been merged, feel free to reopen if you think it needs further discussion
